### PR TITLE
Refactor electrochemical reactions

### LINKED ISSUE #139: Refactor Electrochemical Reactions
**Abstract**

While Cantera/cantera#1181 creates the core infrastructure for the revised `InterfaceKinetics` (via #87), electrochemical reaction (rates) were carved out as they require additional thought (e.g. see #38 and #44; tagging @decaluwe). 

This feature request is created as a reminder that a stop-gap solution needs to be in place for 2.6 as both legacy CTI/XML input as well as the associated legacy framework are slated for immediate removal after the 2.6 release.

### PR DESCRIPTION
**Pull request by ischoegl**
**Changes proposed in this pull request**

- Add ability to evaluate charge-transfer reactions to existing `InterfaceRate<>` and `StickingRate<>` templates
- The existing test suite (specifically `test_kinetics.py::TestSofcKinetics`) covers all changes in detail

**If applicable, fill in the issue number this pull request is fixing**

Closes Cantera/enhancements#139, addresses Cantera/enhancements#142

**If applicable, provide an example illustrating new features this pull request is introducing**

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

**Comment by ischoegl** [comment]
@speth / @decaluwe ... this is ready for review.

**Comment by ischoegl** [comment]
PS: also referencing Cantera/enhancements#38. FWIW, if we can rule out electrochemical 'sticking' formulations (which I currently left in), it would be very easy to introduce a third `ElectrochemicalRate<>` template (in analogy to the already existing `InterfaceRate<>` and `StickingRate<>` templates).

File: include/cantera/kinetics/Reaction.h [hunk 1]
```diff
@@ -118,6 +118,10 @@ class Reaction
     //! @param kin  Kinetics object
     bool checkSpecies(const Kinetics& kin) const;
 
+    //! Check whether reaction uses electrochemistry
+    //! @param kin  Kinetics object
+    bool checkElectrochemistry(const Kinetics& kin) const;
```
Comment (speth) [inline include/cantera/kinetics/Reaction.h]: Perhaps `usesElectrochemistry`, for consistency with the method added to both `CoverageBase` and the Python interface?

File: src/kinetics/Reaction.cpp [hunk 2]
```diff
@@ -64,9 +64,10 @@ Reaction::Reaction(const AnyMap& node, const Kinetics& kin)
         if (nDim == 3) {
             setRate(newReactionRate(node, calculateRateCoeffUnits3(kin)));
         } else {
-            // Reaction type is not specified
+            // Instantiate rate object
```
Comment (speth) [inline src/kinetics/Reaction.cpp]: This comment is kind of odd -- a rate object is being instantiated in either branch of this `if` statement.

File: include/cantera/kinetics/InterfaceRate.h [hunk 3]
```diff
@@ -93,6 +111,95 @@ class CoverageBase
             m_ecov += m_ec[item.first] * shared_data.coverages[item.second];
             m_mcov += m_mc[item.first] * shared_data.logCoverages[item.second];
         }
+
+        // Update change in electrical potential energy
+        if (m_chargeTransfer) {
+            m_deltaPotential_RT = 0.;
+            for (const auto& ch : m_netCharges) {
```
Comment (speth) [inline include/cantera/kinetics/InterfaceRate.h]: I'd say that at this point, this method should be moved to the `.cpp` file

File: include/cantera/kinetics/InterfaceRate.h [hunk 4]
```diff
@@ -93,6 +111,95 @@ class CoverageBase
...
+    }
+
+    //! Calculate modifications for the forward reaction rate for interfacial charge
+    //! transfer reactions.
+    /*!
+     *  For reactions that transfer charge across a potential difference, the
+     *  activation energies are modified by the potential difference. The correction
+     *  factor is based on the net electric potential energy change
+     *  \f[
+     *   deltaElectricEnergy = sum_i ( pot_i nu_ij)
```
Comment (speth) [inline include/cantera/kinetics/InterfaceRate.h]: The formatting of this equation could use some work (sorry if this is just being moved from elsewhere)

File: include/cantera/kinetics/InterfaceRate.h [hunk 5]
```diff
@@ -93,6 +111,95 @@ class CoverageBase
...
+     *  When an electrode reaction rate is specified in terms of its exchange current
+     *  density, the correction factor is adjusted to the standard reaction rate
+     *  constant form and units. Specifically, this converts a reaction rate constant
+     *  that was specified in units of A/m2 to kmol/m2/s.
+     *
+     *  @warning  The updated calculation of voltage corrections is an experimental
+     *      part of the %Cantera API and may be changed or removed without notice.
+     */
+    //
+    double voltageCorrection() const {
```
Comment (speth) [inline include/cantera/kinetics/InterfaceRate.h:117]: There's a bonus `//` comment line here.

File: include/cantera/kinetics/InterfaceRate.h [hunk 6]
```diff
@@ -93,6 +111,95 @@ class CoverageBase
...
+        return correction;
+    }
+
+    //! Boolean indicating whether rate uses electrochemistry
+    /*!
+     *  If this is true, the Butler-Volmer correction is applied
+     *  to the forward reaction rate.
+     *
+     *    fac = exp ( - beta * (delta_phi))
+     */
```
Comment (speth) [inline include/cantera/kinetics/InterfaceRate.h:149]: This equation formatting could be updated.

File: include/cantera/kinetics/InterfaceRate.h [hunk 7]
```diff
@@ -440,7 +550,11 @@ class StickingRate : public RateType, public StickingCoverage
     double evalFromStruct(const DataType& shared_data) const {
         double out = RateType::evalRate(shared_data.logT, shared_data.recipT) *
             std::exp(std::log(10.0) * m_acov - m_ecov * shared_data.recipT + m_mcov);
-
+        if (m_chargeTransfer) {
+            // the physical interpretation of a sticking charge transfer reaction
+            // remains to be resolved.
```
Comment (speth) [inline include/cantera/kinetics/InterfaceRate.h]: Is this worth marking with `@todo`? That way, it will turn up in searches for that (even if comments within methods don't get added to the HTML list of todo items)

**Review by speth — CHANGES_REQUESTED** [review]
Thanks, @ischoegl! This looks really good. I'm encouraged by seeing all of the deprecated methods in `InterfaceKinetics`, as a sign of how much simplification all of this refactoring will eventually bring there.

Beyond the few minor comments below, I had one other thought which is that `CoverageData` and `CoverageBase` have grown to encompass quite a bit more than coverage dependence for the rate. If there's a good reason for not breaking the electrochemical specialization out into a derived class, I'm wondering if there are at least more descriptive names for these classes that we could use.

**Comment by ischoegl** [comment]
@speth .. thank you for the prompt review! Before going over the PR, I wanted to respond to your larger question.

> I had one other thought which is that `CoverageData` and `CoverageBase` have grown to encompass quite a bit more than coverage dependence for the rate. If there's a good reason for not breaking the electrochemical specialization out into a derived class, I'm wondering if there are at least more descriptive names for these classes that we could use.

I was struggling with this also. The single-most important reason for not creating a separate specialization was that I didn't want to block electrochemical 'sticking' reactions. If we can rule this out, then a `ElectrochemicalRate<>` template would be the way to go (but I am wary of the odd-ball mechanism that may want to use this formulation, and @decaluwe - in a brief comment - opined that it's technically possible to go that route despite having some serious misgivings). 

I agree that `CoverageBase` may not be the best descriptor any longer. `InterfaceData` and `InterfaceBase` would be the logical nomenclature, even if it the latter sounds somewhat clunky? Suggestions welcome ... :smile:

**Comment by speth** [comment]
> I agree that `CoverageBase` may not be the best descriptor any longer. `InterfaceData` and `InterfaceBase` would be the logical nomenclature, even if it the latter sounds somewhat clunky? Suggestions welcome ... 😄

What about `InterfaceRateBase` in the latter case, so it doesn't sound like it's a base class for `Interface`, which is something else.

File: include/cantera/kinetics/InterfaceRate.h (hunk 4 above)
Comment (ischoegl) [reply include/cantera/kinetics/InterfaceRate.h]: Yup. Copy-pasta ...

**Comment by ischoegl** [comment]
@speth ... done! 

>  I'm encouraged by seeing all of the deprecated methods in `InterfaceKinetics`, as a sign of how much simplification all of this refactoring will eventually bring there.

It was actually surprising to see that the implementation is actually not that complex - it was just scattered all over and hard to follow.

**Review by speth — APPROVED** [review]
Thanks again, @ischoegl - this all looks good to me.
